core/runtime/ops_rust_to_v8.rs

```diff
@@ -0,0 +1,349 @@
+}
+
+/// Places a value in a `v8::ReturnValue`, non-allocating.
+pub trait RustToV8RetVal<'a>: RustToV8<'a> {
+  fn to_v8_rv(self, rv: &mut v8::ReturnValue<'a>);
+}
+
+/// Convert a value to a `v8::Local`, potentially allocating or failing.
+pub trait RustToV8Fallible<'a> {
+  fn to_v8_fallible(
+    self,
+    scope: &mut v8::HandleScope<'a>,
+  ) -> serde_v8::Result<v8::Local<'a, v8::Value>>;
+}
+
+/// Places a value in a `v8::ReturnValue`, potentially allocating or failing.
+pub trait RustToV8RetValFallible<'a>: RustToV8Fallible<'a> {
+  fn to_v8_rv_fallible(
+    self,
+    scope: &mut v8::HandleScope<'a>,
+    rv: &mut v8::ReturnValue<'a>,
+  ) -> serde_v8::Result<()>;
+}
+
+impl<'a, T> RustToV8<'a> for Option<T>
+where
+  T: RustToV8<'a>,
+{
+  fn to_v8(self, scope: &mut v8::HandleScope<'a>) -> v8::Local<'a, v8::Value> {
+    if let Some(value) = self {
+      value.to_v8(scope)
+    } else {
+      v8::null(scope).into()
+    }
+  }
+}
+
+impl<'a, T> RustToV8Fallible<'a> for Option<T>
+where
+  T: RustToV8Fallible<'a>,
+{
+  fn to_v8_fallible(
+    self,
+    scope: &mut v8::HandleScope<'a>,
+  ) -> serde_v8::Result<v8::Local<'a, v8::Value>> {
+    if let Some(value) = self {
+      value.to_v8_fallible(scope)
+    } else {
+      Ok(v8::null(scope).into())
+    }
+  }
+}
+
+/// Transparent marker struct that can be used to alter the conversion in various ways. The
+/// main uses of this are to mark something as being `serde`-serializable. This may also be
+/// useful to alter the serialization of various types. For example, a `NumberMarker` and
+/// `BigIntMarker` could control the serialization of `i64`/`u64`/`f64` as `v8::Number` or
+/// `v8::BigInt` objects.
+#[repr(transparent)]
+pub struct RustToV8Marker<M, T>(T, PhantomData<M>);
+impl<M, T> From<T> for RustToV8Marker<M, T> {
+  fn from(value: T) -> Self {
+    RustToV8Marker(value, PhantomData)
+  }
+}
+
+/// This struct should be serialized using `serde`.
+pub struct SerdeMarker;
+
+/// Implements a Rust-to-v8 conversion that cannot fail. Multiple types may be specified
+/// to apply the same implementation.
+///
+/// ```no_compile
+/// to_v8!(bool: |value, scope| v8::Boolean::new(scope, value as _));
+/// ```
+macro_rules! to_v8 {
+  (( $( $ty:ty ),+ ) : |$value:ident, $scope:ident| $block:expr) => {
+    $(
+      impl <'a> RustToV8<'a> for $ty {
+        #[inline(always)]
+        fn to_v8(self, scope: &mut v8::HandleScope<'a>) -> v8::Local<'a, v8::Value> {
+          let $value = self;
+          let $scope = scope;
+          v8::Local::<v8::Value>::from($block)
+        }
+      }
+    )+
+  };
+  ($ty:ty : |$value:ident, $scope:ident| $block:expr) => {
+    to_v8!(( $ty ) : |$value, $scope| $block);
+  };
+}
+
+/// Implements a Rust-to-v8 conversion that cannot fail. Multiple types may be specified
+/// to apply the same implementation.
+///
+/// ```no_compile
+/// to_v8!(bool: |value, scope| v8::Boolean::new(scope, value as _));
+/// ```
+macro_rules! to_v8_retval {
```

Comment:
Fixed the comments on them.

Reply:
Added some comments to this effect, but the macros just reduce the boilerplate around these traits. The traits are a combination of "convert to v8 object", "store in v8 returnvalue" crossed with "can fail", "cannot fail". So for example, a u32 can be stored in the ReturnValue and cannot fail. A string cannot be stored in the ReturnValue, and can fail (ie: too large).

The proc macro just maps the different calling patterns to the different interfaces and avoids the overhead of fallible checks if type can never fail to convert, and will ideally call the v8::ReturnValue setters if we can (to avoid explicit object handle allocation).